[Feature] compaction quickly for small data import #9791

chenlinzhong opened the pull request:
# Proposed changes

Issue Number: close #9791

## Problem Summary:
A table frequently imports a small amount of data. Doris generates a version for each imported data. If these versions are not merged in time, two problems will occur

- -235: too many versions, exceeding the maximum number of tablet versions, import failed
- query performance: doris adopts the merge on read mechanism, all versions of data will be merged when reading data. Too many versions may affect query performance

For this reason, we hope to merge the imported versions faster. The simplest way is to increase the number of CC pool thread numbers . However, this may lead to too high io of the entire node, because the CC task may select rowsets with a large amount of data

So We hope to merge some rowsets as soon as possible without adding disk IO

Therefore, we choose rowset with small data rowset to merge first. The definition of small data rowset is that the imported rows are smaller than config:: quick compaction_ max_ rows

To achieve this goal
we add a new  thread pool for  small rowset  compaction. This thread pool only selects rowsets with small data rowsets for merging. The time to trigger merging is as follows:

- Import completed: when be receives the push version request,  try to select a small version of rowset. when the number of rowsets of selected exceeds config:: quick_compaction_ min_ rowsets do a compaction
- -235: try to trigger when -235 appears

### config  parameter
| key | desc | default |
| :-----| :---- | :----: |
| config::enable_quick_compaction |  whether to  turn on this feature | false |
| config::quick_compaction_max_rows | less then rows rowset will be selected  | 1000 |
| config::quick_compaction_batch_size | triggered every publish batch size    | 10 |
| config::quick_compaction_min_rowsets | do quick compaction min selected rowsets | 10 |

### Test
#### 1.we create a table, try to see the quick compaction cost of different rows with centern rowsets
```
CREATE TABLE table1 (
event_day date NULL COMMENT "",
siteid int(11) NULL DEFAULT "10" COMMENT "",
citycode smallint(6) NULL COMMENT "",
username varchar(32) NULL DEFAULT "" COMMENT "",
pv bigint(20) SUM NULL DEFAULT "0" COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(event_day, siteid, citycode, username)
COMMENT "OLAP"
DISTRIBUTED BY HASH(username) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
)
```
here is the result
![image](https://user-images.githubusercontent.com/11487604/172193897-17f83d74-6c1a-4a2f-8c9a-ea832789905a.png)



#### 2. test import small data performance with high frequency 

start 30 threads ,every thread import  100 batch, every batch included 500-1000 rows, the result:

![image](https://user-images.githubusercontent.com/11487604/172792079-2cdd9c28-4085-467a-b29b-9823fa1e28db.png)


![image](https://user-images.githubusercontent.com/11487604/172792386-45a02a8e-5531-4f79-9549-ea155dff2735.png)

![image](https://user-images.githubusercontent.com/11487604/172792404-872b2007-65c9-439d-bb33-dbbd3e5d99c8.png)

<img width="931" alt="image" src="https://user-images.githubusercontent.com/11487604/172794576-b88f4ea5-6d10-407e-bf0a-ee27961b7726.png">


![image](https://user-images.githubusercontent.com/11487604/172792541-2001b8d3-d2c5-482d-8333-dc6a9aeb8e89.png)

we can concluded that 
- cc compaction can not compact rowset in time and will stop import process when reach the max versions limits
- quick compaction can effectively handle  small data import  scene and compact rowset in time

#### 3.test one be import max qps with small data  and compare with quick compaction
[without quick compaction]
start 2 threads ,every thread import  100 batch, every batch included 500-1000 rows
![image](https://user-images.githubusercontent.com/11487604/172859078-1a12e6d8-d77d-496d-a8bd-7bb26ca6030e.png)
![image](https://user-images.githubusercontent.com/11487604/172859103-e087074f-8791-4575-9a24-102447b3212d.png)
we can concluded that 

- max_qps is about 22/s,
- average qps is about 10-15/s 
- import will usually break by -235 

[with quick compaction]
start 30 threads ,every thread import  1000 batch, every batch included 500-1000 rows
![image](https://user-images.githubusercontent.com/11487604/172860310-2eac6122-073e-4e00-849c-238c25d77784.png)
![image](https://user-images.githubusercontent.com/11487604/172860338-ff48ba18-204b-4665-90fe-7f86862d4d18.png)
![image](https://user-images.githubusercontent.com/11487604/172860362-c23b36df-9688-4756-84aa-01676aa6bfbd.png)

we can concluded that 
- max_qps is about 32/s
- average qps is 30/s
- import will never break by -235  and very stable
- disk IO remains low level 

### conclusion

- quick compaction can effectively solve -235 problems 
- quick compaction can imporve  import qps more than three times 
- quick compaction will not  cost a lot of extra disk io
- quick compaction is suitable for scenarios with small amount of data import














## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
4. Has document been added or modified: (Yes/No/No Need)
5. Does it need to update dependencies: (Yes/No)
6. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

chenlinzhong committed:
[Feature] compaction quickly for small data import #9791

chenlinzhong committed:
[Feature] compaction quickly for small data import #9791

morningman commented:
Could you provide some test result for this feature?

Gabriel39 commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
 
+//1.小版本rowset
+//2.不改变cp，cp值的更新有cc任务来执行
+//3.找到最大的，连续小版本rowset
```
Use English comments pls

chenlinzhong replied:
> Use English comments pls

ok

chenlinzhong commented:
> Could you provide some test result for this feature?

ok

BiteTheDDDDt commented on be/src/vec/common/cow.h:
```diff
@@ -350,8 +350,8 @@ class COW {
         T& operator*() { return value->assume_mutable_ref(); }
 
-        operator const immutable_ptr<T>&() const { return value; }
-        operator immutable_ptr<T>&() { return value; }
```
Maybe you should update your formatter. https://github.com/apache/incubator-doris/pull/9602

chenlinzhong replied:
ok

BiteTheDDDDt commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
+    int vertion_count = _tablet->version_count();
+    int64_t now = UnixMillis();
+    _tablet->pick_samll_verson_rowsets(&_input_rowsets);
+    if (_input_rowsets.size() >= 5) {
```
Define a constant in place of the magic number

chenlinzhong replied:
ok

BiteTheDDDDt commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
+    if (tablet_state() == TABLET_RUNNING) {
+        for (auto& rs : _rs_version_map) {
+            bool is_delete = version_for_delete_predicate(rs.first);
+            if (rs.first.first > cumulative_layer_point()) {
```
Maybe you can use more `guard clauses` on here.

caiconghui commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
     return st;
 }
 
+Status Compaction::samll_rowsets_compact() {
```
```suggestion
Status Compaction::small_rowsets_compact() {
```

chenlinzhong replied:
fix

yinzhijian commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
     return st;
 }
 
+Status Compaction::samll_rowsets_compact() {
```
small

chenlinzhong replied:
> small

fixed

yinzhijian commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
+//1.小版本rowset
+//2.不改变cp，cp值的更新有cc任务来执行
+//3.找到最大的，连续小版本rowset
+Status Tablet::pick_samll_verson_rowsets(std::vector<RowsetSharedPtr>* input_rowsets) {
```
small_version

chenlinzhong replied:
> small_version

fixed

yinzhijian commented on be/src/olap/olap_server.cpp:
```diff
@@ -72,6 +72,11 @@ Status StorageEngine::start_bg_threads() {
             .build(&_compaction_thread_pool);
 
+    ThreadPoolBuilder("CompactionTaskThreadPool")
+            .set_min_threads(max_thread_num)
```
add a new config max_small_compaction_threads for finer control
?

yinzhijian commented on be/src/olap/compaction.h:
```diff
@@ -48,6 +48,7 @@ class Compaction {
 
     // This is only for http CompactionAction
     Status compact();
+    Status samll_rowsets_compact();
```
If BaseCompaction.samll_rowsets_compact() is called, what is its behavior?

Gabriel39 commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
+        return Status::OLAPInternalError(OLAP_ERR_CE_TRY_CE_LOCK_ERROR);
+    }
+    _input_rowsets.clear();
+    int vertion_count = _tablet->version_count();
```
```suggestion
    int version_count = _tablet->version_count();
```

chenlinzhong replied:
> 

fixed

Gabriel39 commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,25 @@ Status Compaction::execute_compact() {
+    _tablet->pick_samll_verson_rowsets(&_input_rowsets);
+    if (_input_rowsets.size() >= 5) {
+        do_compaction(0);
+        LOG(INFO) << "samll_rowsets_compact,before_versions:" << vertion_count
```
```suggestion
        LOG(INFO) << "samll_rowsets_compact,before_versions:" << version_count
```

chenlinzhong replied:
> 

fixed

Gabriel39 commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
+    if (max_rows <= 0) return Status::OK();
+    std::vector<std::vector<RowsetSharedPtr>> samll_version_rowsets(max_series_num);
+    int idx = 0;
+    bool is_bad = false;
```
Unused variable?

Gabriel39 commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
+            if (samll_version_rowsets[i].size() > samll_version_rowsets[max_idx].size()) {
+                max_idx = i;
+            }
+            result = samll_version_rowsets[max_idx];
```
seems you should move this line into the above `if` code block

Gabriel39 commented on be/src/olap/tablet.cpp:
```diff
@@ -839,10 +839,53 @@ void Tablet::calculate_cumulative_point() {
+                }
+            }
+        }
+        int max_idx = samll_version_rowsets.size() > 0 ? 0 : -1;
```
Just set `max_idx` to 0

Gabriel39 commented on be/src/olap/delta_writer.cpp:
```diff
@@ -98,6 +100,8 @@ Status DeltaWriter::init() {
     // check tablet version number
     if (_tablet->version_count() > config::max_tablet_version_num) {
+        //trigger samll compaction
+        StorageEngine::instance()->submit_samll_compaction_task(_tablet);
```
should we check whether to pass this check after small compaction?

chenlinzhong replied:
> should we check whether to pass this check after small compaction?

Small version merging is asynchronous, which is unlikely to be completed in a short time, so it is not necessary to check here

Gabriel39 commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -717,7 +719,18 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+            for (int i = 0; i < succ_tablet_ids.size(); i++) {
+                TabletSharedPtr tablet =
+                        StorageEngine::instance()->tablet_manager()->get_tablet(succ_tablet_ids[i]);
+                if (tablet != nullptr) {
```
should we always do small compaction after PublishVersionTask?

chenlinzhong replied:
> should we always do small compaction after PublishVersionTask?

 not necessary ，changed to  batch submit

yangzhg commented on be/src/common/config.h:
```diff
@@ -743,6 +743,9 @@ CONF_Int32(object_pool_buffer_size, "100");
 CONF_Int32(parquet_reader_max_buffer_size, "50");
 
+// For continuous versions that rows less than small_version_max_rows will  trigger compaction quickly
+CONF_Int32(small_version_max_rows, "1000");
```
You'd better to add a config to turn on/off this feature

chenlinzhong replied:
> You'd better to add a config to turn on/off this feature

set 0 means turn off this feature

yangzhg commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -717,7 +719,18 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
-            LOG(INFO) << "publish_version success. signature:" << agent_task_req.signature;
+            for (int i = 0; i < succ_tablet_ids.size(); i++) {
+                TabletSharedPtr tablet =
+                        StorageEngine::instance()->tablet_manager()->get_tablet(succ_tablet_ids[i]);
```
It's better to submit small version compaction in batch, for example  ervery 20 small verison

chenlinzhong replied:
> It's better to submit small version compaction in batch, for example ervery 20 small verison

ok ,done

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_rowset_rows  default 1000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_rowset_rows  default 1000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_rowset_rows  default 1000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_rowset_rows  default 1000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_rowset_rows  default 1000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_max_rows  default 10000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_max_rows  default 10000

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_max_rows  default 10000

yangzhg commented on be/src/common/config.h:
```diff
@@ -300,6 +300,9 @@ CONF_mInt32(convert_rowset_thread_num, "0");
 CONF_mInt32(scan_alpha_rowset_min_interval_sec, "3");
 
+// This config can be set to limit thread number in  smallcompaction thread pool.
+CONF_mInt32(small_compaction_max_threads, "10");
```
use quick compaction is better

yangzhg commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,68 @@ Status Compaction::execute_compact() {
+    int64_t now = UnixMillis();
+    int64_t permits = 0;
+    _tablet->pick_small_verson_rowsets(&_input_rowsets, &permits);
+    std::string input_ver = "";
```
use sstream, and Version has implement operator <<

yangzhg commented on be/src/olap/compaction.cpp:
```diff
@@ -55,6 +55,68 @@ Status Compaction::execute_compact() {
+
+    _input_rowsets.clear();
+    int version_count = _tablet->version_count();
+    int64_t now = UnixMillis();
```
you can use MonotonicStopWatch

yangzhg commented on be/src/olap/tablet.cpp:
```diff
@@ -870,10 +870,56 @@ void Tablet::calculate_cumulative_point() {
+    }
+    int max_series_num = 1000;
+    int max_rows = config::small_compaction_max_rows;
+    if (max_rows <= 0) return Status::OK();
```
```suggestion
    if (max_rows <= 0) { return Status::OK(); }
```

yangzhg commented:
Can you provide some test data about how often this feature can be imported with and without it

caiconghui commented:
will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

chenlinzhong commented:
> will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

maybe， because  small compaction use the same lock with cc compaction, so small compaction may stop  cc compaction for one tablet at this round , but small compaction finished very quickly usually less then 1s , cc compaction maybe excute next round

caiconghui commented:
> > will there be any side effect for normal compaction performance, if there are many stream load and with not small data?
> 
> maybe， because small compaction use the same lock with cc compaction, so small compaction may stop cc compaction for one tablet at this round , but small compaction finished very quickly usually less then 1s , cc compaction maybe excute next round

so，it should be optional choice for user to use this optimization, may be better?

chenlinzhong committed:
compaction quickly for small data import #9791
1.merge small versions of rowset as soon as possible to increase the import frequency of small version data
2.small version means that the number of rows is less than config::small_compaction_max_rows  default 10000

yangzhg reviewed (approved):
LGTM

chenlinzhong commented on be/src/olap/tablet.cpp:
```diff
@@ -870,10 +870,59 @@ void Tablet::calculate_cumulative_point() {
+        }
+        for (int i = 0; i < result.size(); i++) {
+            *permits += result[i]->num_segments();
+            input_rowsets->push_back(result[i]);
```
check consecutive  later in  Compaction::quick_rowsets_compact ,if not will give up

chenlinzhong commented:
> will there be any side effect for normal compaction performance, if there are many stream load and with not small data?

I think it will not , the quick compaciton will do nothing if not with samll data
